bpf/bpf_sock.c

```diff
@@ -357,6 +362,15 @@ static __always_inline int __sock4_xlate_fwd(struct bpf_sock_addr *ctx,
 	if (!svc)
 		return -ENXIO;
 
+#if defined(BPF_HAVE_PERF_EVENT_OUTPUT) && defined(TRACE_SOCK_NOTIFY)
+	if (is_defined(BPF_HAVE_CGROUP_ID))
+		cgroup_id = get_current_cgroup_id();
+
+	send_trace_sock_notify4(ctx_full, XLATE_PRE_DIRECTION_FWD, dst_ip,
+				bpf_ntohs(dst_port), sock_cookie, cgroup_id,
+				ctx_full->protocol);
+#endif
```

Comment:
Bit too much ifdef all around the code here. Please hide sock cookie and cgroup_id retrival inside send_trace_sock_notify{4,6}, so this becomes just a simple function call in the main code.

Also why the double BPF_HAVE_PERF_EVENT_OUTPUT && TRACE_SOCK_NOTIFY?

Can't the agent probe for the helper and only then if available enable TRACE_SOCK_NOTIFY so you can simplify all the ifdefs?

Same with BPF_HAVE_CGROUP_ID, I presume the thought is to still support this infra even when get_current_cgroup_id is not available, right (instead of making it dependent on it)?

nit: Do we need the extra ctx_full->protocol, can't we retrieve it from inside send_trace_sock_notify{4,6} when we pass th ctx_full in there?

Otherwise rest of bpf/ looks good though.